File: duplex/cpp/tests/websocket.cpp
#include <nightseam/duplex/ws.hpp>
#include <future>
#include <iostream>

using namespace nightseam::duplex;
using namespace std::chrono_literals;

void require(bool ok, const char* reason) { if (!ok) throw std::runtime_error(reason); }

void exchange() {
    ws::Listener listener("127.0.0.1", 0, {.subprotocols = {"one", "two"}});
    auto client = ws::dial(listener.url(), {.subprotocols = {"two"}});
    auto server = listener.accept(Wait::after(2s));
    require(client.subprotocol == "two" && server.subprotocol == "two", "subprotocol negotiation changed");
    client.conn->send({Kind::text, "first"});
    client.conn->send({Kind::binary, std::string("\0\xff", 2)});
    auto first = server.conn->receive(Wait::after(2s));
    auto binary = server.conn->receive(Wait::after(2s));
    require(first.kind == Kind::text && first.data == "first", "text frame changed");
    require(binary.kind == Kind::binary && binary.data == std::string("\0\xff", 2), "binary frame changed");
    server.conn->send({Kind::text, "reverse"});
    require(client.conn->receive(Wait::after(2s)).data == "reverse", "reverse frame changed");
    auto closed = std::async(std::launch::async, [&] {
        try { server.conn->receive(Wait::after(2s)); throw std::runtime_error("close was not delivered"); }
        catch (const CloseError& error) { require(error.code == 4012 && error.reason == "finished", "close reason changed"); }
    });
    client.conn->close(4012, "finished", Wait::after(2s));
    closed.get();

    // The listener still accepts a second connection and negotiates nothing
    // when that client offers nothing, despite the server's advertised list.
    auto plain = ws::dial(listener.url());
    auto accepted = listener.accept(Wait::after(2s));
    require(plain.subprotocol.empty() && accepted.subprotocol.empty(), "default subprotocol was invented");
    plain.conn->abort();
    try { accepted.conn->receive(Wait::after(2s)); throw std::runtime_error("abort did not end the remote"); }
    catch (const CloseError& error) { require(error.code == 1006, "abort was not abnormal closure"); }
}

void limit_and_shutdown() {
    ws::Listener listener("127.0.0.1", 0, {.max_frame_bytes = 3});
    auto client = ws::dial(listener.url());
    auto server = listener.accept(Wait::after(2s));
    client.conn->send({Kind::text, "\xf0\x9f\x98\x80"});
    auto ack = std::async(std::launch::async, [&] { try { client.conn->receive(Wait::after(2s)); } catch (...) {} });
    try { server.conn->receive(Wait::after(2s)); throw std::runtime_error("oversized UTF-8 frame arrived"); }
    catch (const FrameTooLarge&) {}
    ack.get();
    auto accepting = std::async(std::launch::async, [&] {
        try { listener.accept(); throw std::runtime_error("closed listener accepted"); }
        catch (const Closed&) {}
    });
    listener.close();
    require(accepting.wait_for(1s) == std::future_status::ready, "listener close left accept waiting");
    accepting.get();
}

void cancellation_and_backpressure() {
    ws::Listener listener;
    auto client = ws::dial(listener.url());
    auto server = listener.accept(Wait::after(2s));
    std::stop_source cancel;
    auto receiving = std::async(std::launch::async, [&] {
        try {
            server.conn->receive({cancel.get_token()});
            throw std::runtime_error("cancelled receive unexpectedly completed");
        } catch (const Cancelled&) {}
    });
    cancel.request_stop();
    require(receiving.wait_for(1s) == std::future_status::ready, "receive cancellation did not settle");
    receiving.get();
    server.conn->abort();
    client.conn->abort();

    auto sender = ws::dial(listener.url());
    auto stopped = listener.accept(Wait::after(2s));
    const Frame large{Kind::binary, std::string(256 << 10, 'x')};
    auto deadline = Wait::after(1s);
    bool paced = false;
    for (int n = 0; n != 400; ++n) {
        try { sender.conn->send(large, deadline); }
        catch (const DeadlineExceeded&) { paced = true; break; }
    }
    require(paced, "WebSocket buffered an unbounded stalled send");
    sender.conn->abort();
    stopped.conn->abort();
}

int main() {
    try { exchange(); limit_and_shutdown(); cancellation_and_backpressure(); }
    catch (const std::exception& error) { std::cerr << error.what() << '\n'; return 1; }
}

File: duplex/rs/tests/websocket.rs
use nightseam_duplex::{Frame, SharedConnection, ws};
use std::time::Duration;
use tokio::time::timeout;

async fn pair(
    limit: usize,
    offered: &[&str],
    selected: &[&str],
) -> (SharedConnection, SharedConnection) {
    let listener = ws::Listener::bind(
        "127.0.0.1:0",
        limit,
        selected.iter().map(|s| s.to_string()).collect(),
    )
    .await
    .unwrap();
    let url = listener.url().unwrap();
    let accept = tokio::spawn(async move { listener.accept().await.unwrap() });
    let client = ws::dial(&url, limit, offered).await.unwrap();
    (client, accept.await.unwrap())
}

#[tokio::test]
async fn websocket_preserves_frames_and_close() {
    let (a, b) = pair(1024, &[], &[]).await;
    assert_eq!(a.subprotocol(), "");
    assert_eq!(b.subprotocol(), "");
    for frame in [
        Frame::Text(vec![]),
        Frame::Text("hé😀".as_bytes().to_vec()),
        Frame::Binary((0..=255).collect()),
    ] {
        a.send(frame.clone()).await.unwrap();
        assert_eq!(b.receive().await.unwrap(), frame);
        b.send(frame.clone()).await.unwrap();
        assert_eq!(a.receive().await.unwrap(), frame);
    }
    for i in 0..64 {
        a.send(Frame::Text(format!("frame {i}").into_bytes()))
            .await
            .unwrap();
        assert_eq!(
            b.receive().await.unwrap(),
            Frame::Text(format!("frame {i}").into_bytes())
        );
    }
    a.send(Frame::Text(b"last".to_vec())).await.unwrap();
    timeout(Duration::from_secs(2), a.close(4001, "finished"))
        .await
        .unwrap()
        .unwrap();
    assert_eq!(b.receive().await.unwrap(), Frame::Text(b"last".to_vec()));
    let ended = b.receive().await.unwrap_err();
    assert_eq!((ended.code, ended.reason.as_str()), (4001, "finished"));
    assert!(a.receive().await.is_err());
    assert!(a.send(Frame::Binary(vec![])).await.is_err());
}

#[tokio::test]
async fn selection_uses_server_preference_and_abort_closes_the_socket() {
    let (a, b) = pair(1024, &["one", "two"], &["two", "one"]).await;
    assert_eq!(a.subprotocol(), "two");
    assert_eq!(b.subprotocol(), "two");
    a.abort();
    assert_eq!(
        timeout(Duration::from_secs(2), b.receive())
            .await
            .unwrap()
            .unwrap_err()
            .code,
        1006
    );
}

#[tokio::test]
async fn receive_limit_and_slow_reader_do_not_hide_an_unbounded_queue() {
    let (a, b) = pair(2, &[], &[]).await;
    a.send(Frame::Binary(vec![0; 3])).await.unwrap();
    assert!(
        timeout(Duration::from_secs(2), b.receive())
            .await
            .unwrap()
            .is_err()
    );
    assert!(b.receive().await.is_err());
    let (a, b) = pair(1 << 20, &[], &[]).await;
    let paced = timeout(Duration::from_millis(200), async {
        for _ in 0..400 {
            a.send(Frame::Binary(vec![1; 256 << 10])).await.unwrap();
        }
    })
    .await;
    assert!(
        paced.is_err(),
        "100 MiB accepted while the reader was stopped"
    );
    a.abort();
    b.abort();
}
